File: frontend/src/optical/kinds/polarizer/physics.ts
/**
 * Polarizer PhysicsOp — ideal linear polarizer with transmission axis
 * at angle `transmissionAxisDegBeamLocal` from beam-local +s.
 *
 * Op name: `jones_polarizer`
 * Kind:    `polarizer`
 *
 * Math:
 *   J(θ) = [ c²    cs  ]      where c = cos θ, s = sin θ
 *          [ cs    s²  ]
 *   jones_out = J · jones_in
 *
 *   Power update: out.power = in.power × |jones_out|² / |jones_in|²
 *   Jones is then renormalized to unit magnitude (so |jones|² ≡ 1 always).
 *
 * Convention:
 *   - Input ray.jones is in beam-local s/p frame (the ray tracer is
 *     responsible for rotating into this frame before invoking the op).
 *   - `transmissionAxisDegBeamLocal = 0` ⇒ transmits +s component (E_s)
 *     and blocks +p (E_p) — Malus law E ∝ cos(θ).
 *
 * Chief ray propagates face_in → face_out along current direction.
 */

import {
  type BeamRay,
  vec3Add,
  vec3Distance,
  vec3Scale,
} from "../../beam-ray";
import { type Complex, cAdd, cMul } from "../../fiber/gaussian";
import {
  type PhysicsOp,
  type PhysicsOpContext,
  registerKind,
} from "../../registry";

function realC(x: number): Complex {
  return { re: x, im: 0 };
}

/** Apply 2×2 real Jones matrix (linear polarizer) to a complex Jones vector. */
export function applyLinearPolarizer(
  jones: [Complex, Complex],
  thetaRad: number,
): [Complex, Complex] {
  const c = Math.cos(thetaRad);
  const s = Math.sin(thetaRad);
  const cc = realC(c * c);
  const cs = realC(c * s);
  const ss = realC(s * s);
  const [Es, Ep] = jones;
  return [
    cAdd(cMul(cc, Es), cMul(cs, Ep)),
    cAdd(cMul(cs, Es), cMul(ss, Ep)),
  ];
}

function jonesMagnitudeSquared(jones: [Complex, Complex]): number {
  const [Es, Ep] = jones;
  return Es.re * Es.re + Es.im * Es.im + Ep.re * Ep.re + Ep.im * Ep.im;
}

export const jonesPolarizerOp: PhysicsOp = (
  rayIn: BeamRay,
  ctx: PhysicsOpContext,
): BeamRay[] => {
  const axisDeg = (
    ctx.params.transmissionAxisDegBeamLocal
    ?? ctx.params.transmissionAxisDegBodyLocal
    ?? 0
  ) as number;
  const theta = (axisDeg * Math.PI) / 180;

  const jonesProjected = applyLinearPolarizer(rayIn.jones, theta);
  const magSqIn = jonesMagnitudeSquared(rayIn.jones);
  const magSqOut = jonesMagnitudeSquared(jonesProjected);

  // Transmittance (relative to input magnitude). Guard against zero input.
  const t = magSqIn > 1e-30 ? magSqOut / magSqIn : 0;
  const newPower = rayIn.powerMw * t;

  // Renormalize Jones to unit magnitude (or keep as input if blocked).
  let newJones: [Complex, Complex];
  if (magSqOut > 1e-30) {
    const norm = 1 / Math.sqrt(magSqOut);
    newJones = [
      { re: jonesProjected[0].re * norm, im: jonesProjected[0].im * norm },
      { re: jonesProjected[1].re * norm, im: jonesProjected[1].im * norm },
    ];
  } else {
    // Beam fully blocked — Jones direction undefined; keep input for safety.
    newJones = rayIn.jones;
  }

  // Chief ray propagates face_in → face_out along current direction.
  const thicknessMm = vec3Distance(
    ctx.faceIn.positionMmBodyLocal,
    ctx.faceOut.positionMmBodyLocal,
  );
  const newOrigin = vec3Add(rayIn.origin, vec3Scale(rayIn.direction, thicknessMm));

  return [{
    ...rayIn,
    origin: newOrigin,
    jones: newJones,
    powerMw: newPower,
    pathLengthMm: rayIn.pathLengthMm + thicknessMm,
  }];
};

registerKind("polarizer", {
  ops: {
    jones_polarizer: jonesPolarizerOp,
  },
  needsAperture: true,
});

// Variant ops (Glan-Laser PBS-like rejection) register themselves under
// the polarizer kind via registerOps when imported. The eager-import in
// kinds/__init__ guarantees ordering: polarizer first, then glan-laser.

File: backend/app/optical/kinds/polarizer/physics.py
"""Polarizer PhysicsOp (Python mirror of frontend kinds/polarizer/physics.ts).

Op name: jones_polarizer
Kind: polarizer
Required params: transmissionAxisDegBeamLocal (defaults 0)
"""

from __future__ import annotations

import math

from app.optical.beam_ray import BeamRay, vec3_distance
from app.optical.registry import (
    KindEntry,
    PhysicsOpContext,
    register_kind,
)


def apply_linear_polarizer(
    jones: tuple[complex, complex], theta_rad: float
) -> tuple[complex, complex]:
    """J(theta) = [[c², cs], [cs, s²]]  applied to (E_s, E_p)."""
    c = math.cos(theta_rad)
    s = math.sin(theta_rad)
    cc, cs, ss = c * c, c * s, s * s
    e_s, e_p = jones
    return (cc * e_s + cs * e_p, cs * e_s + ss * e_p)


def _jones_mag_sq(jones: tuple[complex, complex]) -> float:
    e_s, e_p = jones
    return (e_s.real * e_s.real + e_s.imag * e_s.imag
            + e_p.real * e_p.real + e_p.imag * e_p.imag)


def jones_polarizer_op(ray_in: BeamRay, ctx: PhysicsOpContext) -> list[BeamRay]:
    axis_deg = float(
        ctx.params.get("transmissionAxisDegBeamLocal")
        or ctx.params.get("transmissionAxisDegBodyLocal")
        or 0.0
    )
    theta = math.radians(axis_deg)

    jones_proj = apply_linear_polarizer(ray_in.jones, theta)
    mag_sq_in = _jones_mag_sq(ray_in.jones)
    mag_sq_out = _jones_mag_sq(jones_proj)

    t = mag_sq_out / mag_sq_in if mag_sq_in > 1e-30 else 0.0
    new_power = ray_in.power_mw * t

    if mag_sq_out > 1e-30:
        norm = 1.0 / math.sqrt(mag_sq_out)
        new_jones = (jones_proj[0] * norm, jones_proj[1] * norm)
    else:
        new_jones = ray_in.jones

    thickness_mm = vec3_distance(
        ctx.face_in.position_mm_body_local,
        ctx.face_out.position_mm_body_local,
    )
    new_origin = ray_in.origin + ray_in.direction * thickness_mm

    return [ray_in.replaced(
        origin=new_origin,
        jones=new_jones,
        power_mw=new_power,
        path_length_mm=ray_in.path_length_mm + thickness_mm,
    )]


register_kind("polarizer", KindEntry(
    ops={"jones_polarizer": jones_polarizer_op},
    needs_aperture=True,
))
